Ghost constraint improvements

juliand665 opened the pull request:
I reworked ghost constraint parsing to directly use `syn` for the trait bounds and comma, only then handing the rest over to Prusti's parser. Additionally, I extended them to handle `pure` flags (in addition to the status quo of `requires` and `ensures`). The effects are as follows:
- Any trait bounds that are valid rust are now inherently valid in ghost constraints.
- The separating comma between bounds & specs no longer requires a space if preceded by a `>`.
- Ghost constraints can now specify `pure`ness.
- The `ghost_constraint` proc macro is now correctly a no-op when compiling without prusti, rather than removing the function entirely.

Some considerations for code review:
- I just kinda threw in the logic with the rest of it in preparser.rs; perhaps splitting it (or other parts) off into its own file would make sense? It's a pretty long file at this point.
- I wasn't sure how best to translate the existing unit tests now that we no longer represent the traits as an unparsed `TokenStream`. I ended up providing the strings as `quote!` streams instead, but it feels like that's not really testing anything of value anymore (hence also why I dropped it in the new `tuple_generics` test).
- The errors reported for invalid syntax are better in some ways and worse in other ways compared to the status quo. What I most miss is that the old errors often provided an example of how to structure your ghost constraint—perhaps it would make sense to somehow attach this to whatever error syn outputs?

juliand665 committed:
allow pure as ghost constraint

also tried my hand at streamlining some of the parsing, not that happy with it yet though

juliand665 committed:
rely on syn to parse trait bounds

this fixes weirdness involving commas:
- no longer requires a space before the comma separating the bounds from the specs
- now correctly handles tuples (e.g. `ghost_constraint(F: FnOnceSpec<(i32,)>, […])`

juliand665 committed:
update tests

juliand665 committed:
tiny rename

juliand665 committed:
expand test suite

now tests parsing pure attribute and some tuple types that caused trouble before

juliand665 committed:
update standalone tests

hopefully this is all of them; i can't run them locally right now

juliand665 committed:
remove unnecessary into()

not sure why i thought this was necessary; it's already a result

Aurel300 commented on prusti-specs/src/lib.rs:
```diff
@@ -238,6 +239,10 @@ fn generate_for_pure(attr: TokenStream, item: &untyped::AnyFnItem) -> GeneratedR
+    do_generate_for_pure(item)
+}
+
+fn do_generate_for_pure(item: &untyped::AnyFnItem) -> GeneratedResult {
```
This is not an amazing name, perhaps add a one-line comment explaining that this one is usable from other parts of `prusti-specs` and that it doesn't receive or check the attribute tokens?

juliand665 replied:
Good catch! In any other language, I would have just overloaded it, but coming up with a meaningful name is hard. Maybe we could incorporate the word `unchecked` somehow?

juliand665 replied:
Actually, what if we get rid of this and instead call `generate_spec_and_assertions` directly from `ghost_constraints::generate`? We'd could either remap `NestedSpec` to `SpecAttributeKind` or directly parse as the latter to get rid of the current (minor) duplicate parsing logic, though this would affect `|=` as well…

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -81,6 +80,7 @@ fn error<T>(span: Span, msg: &str) -> syn::Result<T> {
 struct PrustiTokenStream {
     tokens: VecDeque<PrustiToken>,
     source_span: Span,
+    // TODO: can we somehow update the span after popping stuff?
```
Would this get us more precise error reporting? I think the span here is for cases where there are no tokens but we still want a span (of the enclosing group, for example).

juliand665 replied:
For example when you use the new `parse_rest` function and it encounters additional tokens past what it needed to parse, I was expecting the span to just cover those unconsumed tokens at that point, and I feel like that would be generally useful. For cases where there are no tokens, I'm not sure the span of "everything parsed as prusti tokens" is more useful than just `Span::call_site()` and such.

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -161,6 +161,17 @@ impl PrustiTokenStream {
+    {
+        let result = f(&mut self)?;
+        if !self.is_empty() {
+            return error(self.source_span, "unexpected extra tokens");
```
You can use the span of the first token, since you know there is at least one left.

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -410,32 +391,29 @@ impl PrustiTokenStream {
+    fn pop_single_nested_spec(&mut self) -> syn::Result<NestedSpec<Self>> {
+        let first = self.tokens.pop_front().map_or_else(|| {
+            error(self.source_span, "expected nested spec")
+        }, Ok)?;
```
A version of the `error` helper method which only returns the `syn::parse::Error` might be nice. Then this can be an `ok_or_else`, and the errors in `expr_bp` would be a bit nicer.

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -508,18 +485,87 @@ impl PrustiTokenStream {
+    fn as_pts(&self) -> PrustiTokenStream {
+        PrustiTokenStream::new(self.parse().unwrap())
+    }
+}
```
This is a trait just to be able to call it as `.as_pts()`?

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -508,18 +485,87 @@ impl PrustiTokenStream {
+    }
+}
+
+// TODO: is there a better place for this type and its logic?
```
Let's keep it here for now. We don't have so many spec syntax components yet.

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -907,46 +953,64 @@ mod tests {
-            let (bounds, nested_specs) = parse_ghost_constraint(quote!{ T: A, []}).unwrap();
-            assert_eq!(bounds.to_string(), "T : A");
-            assert!(nested_specs.is_empty());
+            // TODO: does it even make sense to compare parsed quotes like this? feels a bit like saying 2 = 2
```
This test is very simple, so the comparison of the trait bound isn't terribly interesting. For the others it makes more sense, and this way it is consistent.

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -907,46 +953,64 @@ mod tests {
+        
+        #[test]
+        fn tuple_generics() {
+            // just check that parsing succeeds
```
Why not `assert!(... .is_ok());`?

juliand665 replied:
Good idea! Not sure why I didn't think of that.

Aurel300 commented on prusti-tests/tests/parse/fail/ghost-constraints/trait-bounds-invalid-macro-arguments-1.rs:
```diff
@@ -1,6 +1,6 @@
 use prusti_contracts::*;
 
-#[ghost_constraint([ //~ ERROR: Invalid use of macro. Two arguments expected (a trait bound `T: A + B` and multiple specifications `[requires(...), ensures(...), ...]`)
+#[ghost_constraint([ //~ ERROR: expected `,`
```
Yeah this error is not great. Ideally we would show an expected form for the attribute with any error emitted, but I don't see a way to attach "notes" to `syn` errors.

juliand665 committed:
compute more precise span for extra tokens error

juliand665 committed:
use assert rather than unwrap

juliand665 committed:
make ghost_constraint not remove functions when prusti isn't present

juliand665 committed:
remove premature abstraction

juliand665 committed:
add note with expected structure

can't make this an actual NOTE but this already helps a lot

juliand665 committed:
work around test env weirdness

juliand665 committed:
update e2e parsing test

too bad the message is coming from trait bound parsing, but hey, the note still helps

juliand665 committed:
add explanation for do_generate_for_pure

juliand665 committed:
add helper to construct syn::Error

had to rename the existing one because i think this name makes more sense and i needed them to stay short

juliand665 committed:
remove discussed TODOs

Aurel300 commented on prusti-tests/tests/parse/fail/ghost-constraints/trait-bounds-illegal-2.rs:
```diff
@@ -2,9 +2,10 @@ use prusti_contracts::*;
+#[ghost_constraint(T: for<'a> A<&'a i32>, [ //~ ERROR: lifetimes are not allowed in ghost constraint trait bounds
     ensures(result > 0)
 ])]
+//~| ERROR: expected a trait bound `T: A + B` and specifications in brackets `[requires(...), ensures(...), pure, ...]`
```
Here the extra error is maybe unnecessary, but we would have to check (by string comparison?) that the emitted error is not just a generic `syn` parse error, right?

juliand665 replied:
I have to specify all errors that the test expects to generate; any additional or missing ones will cause it to fail. The way syn outputs these combined errors is as two separate errors with the same span (closest we can get to a note).

Aurel300 replied:
I understand that but here our error is specifically that we disallow lifetimes, which is an error we emit ourselves, as opposed to a generic `syn` parsing error.

juliand665 replied:
Ah, my bad, I see what you mean now. Yeah, it's kind of silly to attach an error like this in that case. It doesn't do much harm, so I wonder if it justifies the additional complexity of moving down the error to more specific parsing logic—I'll push a commit that does that so we can compare.

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -51,7 +51,11 @@ pub fn parse_prusti_assert_pledge(tokens: TokenStream) -> syn::Result<(TokenStre
-    syn::parse2(tokens)
+    syn::parse2(tokens).map_err(|mut err| {
+        err.combine(syn::Error::new(err.span(),
+            "expected a trait bound `T: A + B` and specifications in brackets `[requires(...), ensures(...), pure, ...]`"));
```
This is better, but still doesn't give an actual example of the syntax. I think the error message should contain the expected format `ghost_constraint(T: A + B + ..., [requires(...), ...])`

juliand665 committed:
provide example in error message

juliand665 committed:
only emit example syntax when deemed relevant

Aurel300 commented on prusti-specs/src/specifications/preparser.rs:
```diff
@@ -410,44 +401,40 @@ impl PrustiTokenStream {
+    fn pop_parenthesized_group(&mut self) -> syn::Result<Self> {
+        match self.tokens.pop_front() {
+            Some(PrustiToken::Group(_span, Delimiter::Parenthesis, box group)) => {
+                Ok(group) // TODO: need to clone()?
```
What is this TODO about?

juliand665 replied:
Not sure; that one was there before I touched the code! I would also expect it to be clear whether or not we need to clone from a type safety standpoint, but I don't know enough about syn to be sure.
